src/libs/Navigation/AppNavigator/ModalStackNavigators.js

```diff
@@ -13,11 +14,12 @@ const defaultSubRouteOptions = {
  * Create a modal stack navigator with an array of sub-screens.
  *
  * @param {Object[]} screens array of screen config objects
+ * @param {React.ReactElement} Provider a wrapper component with context provider to share data between screens
  * @returns {Function}
  */
-function createModalStackNavigator(screens) {
+function createModalStackNavigator(screens, Provider) {
```

Comment:
This seems like a new pattern. Also it's causing a bug: navigating to those screens goes through another url.
1. Click FAB
2. Click Send Money
3. Notice url changes to `/?initial=true&screen=Money_Request&params=...`
4. Then to `/send/new`

Reply:
Hmm, I will check this one tomorrow.

Reply:
Here is what I found:

When we open the money request page, I see in the log the navigation state changes twice, which only should be one for /send/new. Here is the comparison:

| First State (unexpected, random params) | Second State (expected, /send/new) |
| ------------- | ------------- |
| <img width="362" alt="image" src="https://github.com/Expensify/App/assets/50919443/8ff4ea8d-28e4-4094-8b72-4ef741721610"> | <img width="352" alt="image" src="https://github.com/Expensify/App/assets/50919443/686f7878-6c31-49e3-b1f1-43b0cb666224"> |

We can see that the different is the correct state has the `state` property, while the wrong one does not have it. As we wrap the MoneyRequestProvider with onyx, it will return null initially which I guess the cause of the first unexpected navigation state. Removing the `withCurrentUserPersonalDetails` will solve the issue. So, the solution I could think of is to remove the onyx, initialize the currency state with USD, then we get the correct currency in MoneyRequestAmountPage. Here is the diff:

```diff
diff --git a/src/pages/iou/steps/MoneyRequestAmountPage.js b/src/pages/iou/steps/MoneyRequestAmountPage.js
index 1d4f1a001d..dde39efe11 100755
--- a/src/pages/iou/steps/MoneyRequestAmountPage.js
+++ b/src/pages/iou/steps/MoneyRequestAmountPage.js
@@ -22,6 +22,7 @@ import ScreenWrapper from '../../../components/ScreenWrapper';
 import FullPageNotFoundView from '../../../components/BlockingViews/FullPageNotFoundView';
 import withMoneyRequest, {moneyRequestPropTypes} from '../withMoneyRequest';
 import withNavigation from '../../../components/withNavigation';
+import withCurrentUserPersonalDetails from '../../../components/withCurrentUserPersonalDetails';
 import ModalHeader from '../ModalHeader';
 import ONYXKEYS from '../../../ONYXKEYS';
 import personalDetailsPropType from '../../personalDetailsPropType';
@@ -74,10 +75,25 @@ class MoneyRequestAmountPage extends React.Component {
         this.reportID = lodashGet(props.route, 'params.reportID', '');
         this.isEditing = lodashGet(props.route, 'path', '').includes('amount');
 
-        const selectedAmountAsString = props.moneyRequest.amount ? CurrencyUtils.convertToWholeUnit(props.moneyRequest.currency, props.moneyRequest.amount).toString() : '';
+        // Connecting MoneyRequestProvider with Onyx will briefly show some "random" params 
+        // in the URL when opening money request page.
+        // As both currency and participants require data from onyx, we initialize them here.
+        let currency = props.moneyRequest.currency;
+        if (!this.isEditing) {
+            currency = lodashGet(props.currentUserPersonalDetails, 'localCurrencyCode', CONST.CURRENCY.USD);
+            props.moneyRequest.setCurrency(currency);
+
+            // Set the money request participants based on the report participants
+            const participants = ReportUtils.isPolicyExpenseChat(props.report)
+                ? OptionsListUtils.getPolicyExpenseReportOptions(props.report)
+                : OptionsListUtils.getParticipantsOptions(props.report, props.personalDetails);
+            props.moneyRequest.setParticipants(_.map(participants, (participant) => ({...participant, selected: true})));
+        }
+
+        const selectedAmountAsString = props.moneyRequest.amount ? CurrencyUtils.convertToWholeUnit(currency, props.moneyRequest.amount).toString() : '';
         this.state = {
             amount: selectedAmountAsString,
-            selectedCurrencyCode: props.moneyRequest.currency,
+            selectedCurrencyCode: currency,
             shouldUpdateSelection: true,
             selection: {
                 start: selectedAmountAsString.length,
@@ -87,17 +103,10 @@ class MoneyRequestAmountPage extends React.Component {
     }
 
     componentDidMount() {
-        if (this.isEditing) {
-            if (_.isEmpty(this.props.moneyRequest.participants) || this.props.moneyRequest.amount === 0) {
-                Navigation.goBack();
-                Navigation.navigate(ROUTES.getMoneyRequestRoute(this.iouType, this.reportID));
-            }
-        } else {
-            // Set the money request participants based on the report participants
-            const participants = ReportUtils.isPolicyExpenseChat(this.props.report)
-                ? OptionsListUtils.getPolicyExpenseReportOptions(this.props.report)
-                : OptionsListUtils.getParticipantsOptions(this.props.report, this.props.personalDetails);
-            this.props.moneyRequest.setParticipants(_.map(participants, (participant) => ({...participant, selected: true})));
+        // Prevent user to access edit amount page if both participant selection and initial amount step is not completed yet
+        if (this.isEditing && (_.isEmpty(this.props.moneyRequest.participants) || this.props.moneyRequest.amount === 0)) {
+            Navigation.goBack();
+            Navigation.navigate(ROUTES.getMoneyRequestRoute(this.iouType, this.reportID));
         }
 
         this.focusTextInput();
@@ -448,6 +457,7 @@ MoneyRequestAmountPage.defaultProps = defaultProps;
 export default compose(
     withLocalize,
     withNavigation,
+    withCurrentUserPersonalDetails,
     withMoneyRequest,
     withOnyx({
         report: {
diff --git a/src/pages/iou/withMoneyRequest.js b/src/pages/iou/withMoneyRequest.js
index 31b4f4c97a..ce0352ceda 100644
--- a/src/pages/iou/withMoneyRequest.js
+++ b/src/pages/iou/withMoneyRequest.js
@@ -5,6 +5,7 @@ import withCurrentUserPersonalDetails, {withCurrentUserPersonalDetailsDefaultPro
 import * as PersonalDetails from '../../libs/actions/PersonalDetails';
 import optionPropTypes from '../../components/optionPropTypes';
 import useOnNetworkReconnect from '../../components/hooks/useOnNetworkReconnect';
+import CONST from '../../CONST';
 
 const MoneyRequestContext = createContext();
 const moneyRequestPropTypes = PropTypes.shape({
@@ -35,7 +36,7 @@ const moneyRequestPropTypes = PropTypes.shape({
 
 const MoneyRequestProvider = (props) => {
     const [amount, setAmount] = useState(0);
-    const [currency, setCurrency] = useState(props.currentUserPersonalDetails.localCurrencyCode);
+    const [currency, setCurrency] = useState(CONST.CURRENCY.USD);
     const [participants, setParticipants] = useState([]);
     const [comment, setComment] = useState('');
 
@@ -69,7 +70,7 @@ MoneyRequestProvider.defaultProps = {
     ...withCurrentUserPersonalDetailsDefaultProps,
 };
 
-const MoneyRequestProviderWithOnyx = withCurrentUserPersonalDetails(MoneyRequestProvider);
+const MoneyRequestProviderWithOnyx = MoneyRequestProvider;
 
 export default function withMoneyRequest(WrappedComponent) {
     const WithMoneyRequest = forwardRef((props, ref) => (

```

Let me know what you think

Reply:
I would prefer not to go this way. Can we instead change the location of the Provider?

Reply:
If we put it higher in the tree, all pages will be able to access the money request state. We can't move it deeper as that means only one screen will receive the context value. The reason I put it for money request navigator only so only money request related pages can access the state and when we close the page, all of the state is lost.

Reply:
Can you please check if we can do it in another way? e.g. use another context, use withOnyx (initWithStoredValues: false) or any solution as long as we render navigation stack and not `null`.

Reply:
Using `initWithStoredValues: false` will make the data undefined because we don't init the data with the stored values. Still looking for other solution.

Reply:
Here is the alternative:

As we only use current user personal details (which is taken from personal details and session onyx) to initialize the currency initial value, we don't need to wrap the provider with onyx. Instead, we can just get the value on demand in the state initial value. But, where do we get the current user personal details or even the user local currency code? We have 3 options:

1. Add Onyx.connect of personal details and session in withMoneyRequest. Example:
https://github.com/Expensify/App/blob/8180113819532fe5503bc3719ba46ffc566e4ead/src/libs/actions/App.js#L19-L46

2. Create a new util function called `getCurrentUserPersonalDetails` in `PersonalDetailsUtils`.
3. Create a new util function called `getUserLocalCurrencyCode` in `IOUUtils`.

_It would be great if there is a `Onyx.get` function. I guess there are lot of places where we only need the data from Onyx once for initialization but then we wrap it with `withOnyx`, which means the component will unnecessarily re-render even though we don't care about the updated value_

Reply:
We don't use Onyx.connect with components.

Reply:
Technically we use it in AuthScreens (I just realized it) and we connect with both session and personal details which what we needed. So, we can use that and pass the current user personal details (or maybe just the currency?) to the `initialParams` of MoneyRequest screen. Then, we get it from route params in withMoneyRequest.

Btw, there are 2nd and 3rd options too.

Reply:
I guess that would work. But still anti-pattern. Those onyx callbacks are not used with components.

Reply:
This seems to be complex problem :sweat_smile:. I think we are looking for something like this https://stackoverflow.com/questions/65629534/how-to-wrap-multiple-stack-screens-with-context-provider-in-react-navigation-5 but that does not look possible yet.

I think we may have to use the context solution i.e. creating another context provider for onyx. I tested `withPersonalDetails` and it's seems to work.

```js
const [currency, setCurrency] = useState(lodashGet(props.personalDetails, ['admin@abdelhafidh.com', 'localCurrencyCode']), 'USD');
...
const MoneyRequestProviderWithOnyx = withPersonalDetails()(MoneyRequestProvider);
```

So we will have to write another HOC `withCurrentUserPersonalDetails` but it should use context (see [`createOnyxContext`](https://github.com/Expensify/App/blob/main/src/components/createOnyxContext.js))


However the idea of having two `withCurrentUserPersonalDetails` HOCs (one that uses withOnyx directly and one that uses it via a context) is not ideal, so maybe we should remove the first one? that way both  `withPersonalDetails` and `withCurrentUserPersonalDetails` will be similar and both will be provided to the whole app.

Reply:
> That's possible and that's what we do right now

Currently we do Provider > Navigator > Screens. But this seems to cause a problem. I meant to do Navigator >Provider > Screens. But Navigator only accept Screens as children.

___

`getUserLocalCurrencyCode` That looks to be the easiest solution :grinning: But do we follow such pattern anywhere? I mean this will return a value that we get from onyx connect afterall.

Reply:
`getUserLocalCurrencyCode` looks like just a "getter" to the onyx and "do we follow such pattern anywhere?"

I guess the most similar one is this, literally just return the data from Onyx 😄 

https://github.com/Expensify/App/blob/c917c447f225fe3b194a5e03f568cca0de9e5e71/src/libs/actions/PersistedRequests.js#L42-L44

But I don't think we have something like getting the onyx data from some files in a component.

Reply:
I guess that would be a perfect fit in our case since we just need the initial value at render (assuming we will remove the openMoneyRequestModalPage calls).

I think I'm okay with going with that approach. @PauloGasparSv @thienlnam do you have any thoughts here?

Reply:
Could you please summarize what the problem that we're running into is? From a brief glance we're doing an App navigation reboot which may help with problem, but don't have much insight into that

Reply:
To know:
If we [render](https://github.com/Expensify/App/blob/e7521264711339bbdb92363b82ff09a148f3655b/src/libs/Navigation/AppNavigator/AuthScreens.js#L310) a `null` stack navigator, the url would show the routing parameters e.g. `/?initial=true&screen=Money_Request&params=...`.

Problem:
`MoneyRequestModalStackNavigator` uses a Provider that uses Onyx (`withOnyx`). Components that use withOnyx will render null at the very first render.

Solution:
Do not use withOnyx. Instead use Onyx.connect.

What is not clear:
Usually we don't use Onyx.connect to store a value that is meant for direct use by a component (we use withOnyx for that) but in our case the use of Onyx.connect seems quite tempting as it's the easiest solution we found yet.

Question:
Do you think we should go for it, do nothing, something else?

Reply:
IMO, that would be an anti-pattern and a sign that it's being held wrong. Could you please bring to #expensify-open-source?

Reply:
https://expensify.slack.com/archives/C01GTK53T8Q/p1684873343572679

Reply:
Looks like we haven't reach any agreement in Slack. I would say we go with the current plan. I'm also planning to move the initialization of the participants from amount page to the provider, so we will have 2 new function in IOUUtils, 1 for getting the current user local currency code, 1 for get money request participants.

Reply:
Can you bump the thread again, let's wait a couple more days

Reply:
Bumped

Reply:
We reached a conclusion to use Onyx instead. Going to update it!